cranelift/filetests/src/function_runner.rs

```diff
@@ -1,96 +1,381 @@
+    pub fn new(page: Mmap, signature: Signature, trampoline: &'a Trampoline) -> Self {
+        Self {
+            page,
+            signature,
+            trampoline,
+        }
+    }
+
+    /// Return a pointer to the compiled code.
+    pub fn as_ptr(&self) -> *const u8 {
+        self.page.as_ptr()
+    }
+
+    /// Call the [CompiledFunction], passing in [DataValue]s using a compiled [Trampoline].
+    pub fn call(&self, arguments: &[DataValue]) -> Vec<DataValue> {
+        let mut arguments = make_arguments(arguments, &self.signature);
+        let arguments_address = arguments.as_mut_ptr();
+        let function_address = self.as_ptr();
+
+        let callable_trampoline: fn(*const u8, *mut u128) -> () =
+            unsafe { mem::transmute(self.trampoline.as_ptr()) };
+        callable_trampoline(function_address, arguments_address);
+
+        collect_returns(&arguments, &self.signature)
+    }
+}
+
+/// Compile a [Function] to its executable bytes in memory.
+///
+/// This currently returns a [Mmap], a type from an external crate, so we wrap this up before
+/// exposing it in public APIs.
+fn compile(function: Function, isa: &dyn TargetIsa) -> Result<Mmap, CompilationError> {
+    // Set up the context.
+    let mut context = Context::new();
+    context.func = function;
 
-        if func.signature.call_conv != self.isa.default_call_conv() {
-            return Err(String::from(
-                "Functions only run on the host's default calling convention; remove the specified calling convention in the function signature to use the host's default.",
-            ));
+    // Compile and encode the result to machine code.
+    let relocs = &mut NullRelocSink {};
+    let traps = &mut NullTrapSink {};
+    let stackmaps = &mut NullStackmapSink {};
+    let code_info = context.compile(isa)?;
+    let mut code_page = MmapMut::map_anon(code_info.total_size as usize)?;
+
+    unsafe {
+        context.emit_to_memory(isa, code_page.as_mut_ptr(), relocs, traps, stackmaps);
+    };
+
+    let code_page = code_page.make_exec()?;
+    trace!(
+        "Compiled function {} with signature {} at: {:p}",
+        context.func.name,
+        context.func.signature,
+        code_page.as_ptr()
+    );
+
+    Ok(code_page)
+}
+
+/// Build the arguments vector for passing the [DataValue]s into the [Trampoline]. The size of
+/// `u128` used here must match [Trampoline::SLOT_SIZE].
+fn make_arguments(arguments: &[DataValue], signature: &ir::Signature) -> Vec<u128> {
+    assert_eq!(arguments.len(), signature.params.len());
+    let mut values_vec = vec![0; max(signature.params.len(), signature.returns.len())];
+
+    // Store the argument values into `values_vec`.
+    for ((arg, slot), param) in arguments.iter().zip(&mut values_vec).zip(&signature.params) {
+        if arg.ty() != param.value_type {
+            panic!("argument type mismatch");
+        }
+        unsafe {
+            Trampoline::write_value_to(arg, slot);
```

Comment:
Could we just do a write here, using a transmute? It avoids the indirection through `Trampoline::write_value_to`, which is a bit unsettling.

If we did the same for reads, then the need for the `Trampoline` data structure is much lower, and could probably be avoided. Or we could hide the writing and reading into a different new `UnboxedValues` struct type (that'd be a vec<u128>); then the `slot_size` could be an attribute of this struct, a const func returning the size of the values it's storing (and then we can remove a bunch of comments, and have more confidence in the size being sync'd with the storage type). What do you think?

Reply:
I like having the `Trampoline` structure around to know what we are talking about; otherwise we have an `Mmap` floating around that is implicitly a trampoline. But that could be a type alias, I guess. I think your `UnboxedValues` makes sense and I started down this road but I'm not too confident about the transmute stuff--I will ping you on Zulip to see what you think.